rep/publisher.go

```diff
@@ -0,0 +1,36 @@
+package rep
+
+import "sync"
+
+type RetrievalEventPublisher struct {
+	mu          sync.Mutex
+	subscribers map[RetrievalSubscriber]struct{}
+}
+
+func NewEventPublisher() (*RetrievalEventPublisher, error) {
+	return &RetrievalEventPublisher{
+		subscribers: map[RetrievalSubscriber]struct{}{},
```

Comment:
Just realized you're almost surely not going to be able to construct a map this way. When you actually attempt to use FilClient or Autoretrieve as the map key you'll likely get a panic as they aren't comparable types.

If you need a piece of code to model this on, this is the little library I wrote, which is probably too generic but still easy to copy ideas from -- https://github.com/hannahhoward/go-pubsub/blob/master/pubsub.go

Reply:
https://go.dev/ref/spec#Comparison_operators -- for reference on what can and cannot be an map key -- it must be "comparable"